File: 0839-similar-string-groups/0839-similar-string-groups.cpp
class Solution {
    class DSU {
        public:
        vector<int>par,size;
        DSU(int n)
        {par=vector<int>(n);size=vector<int>(n);for(int i=0;i<n;i++){par[i]=i;size[i]=1;}}
        int find(int x){if(x==par[x])return x;return par[x]=find(par[x]);}
        bool is_joined(int a,int b){int pa=find(a);int pb=find(b);
        if(pa==pb)return 1;
        if(size[pa]<=size[pb]){par[pa]=pb;size[pb]+=size[pa];}
        else{par[pb]=pa;size[pa]+=size[pb];}
        return 0;}
    };
public:
    int numSimilarGroups(vector<string>& strs) {
        int n=strs.size();
        DSU dsu(n);
        
        for(int i=0;i<n;i++) {
            for(int j=i+1;j<n;j++) {
                vector<int>v;
                for(int k=0;k<strs[i].length();k++) {
                    if(strs[i][k]!=strs[j][k]) {
                        v.push_back(k);
                    }
                }
                if(v.size()==0||(v.size()==2&&strs[i][v[0]]==strs[j][v[1]]&&strs[i][v[1]]==strs[j][v[0]])) {
                    dsu.is_joined(i,j);
                }
            }
        }
        
        set<int>parents;
        for(int i=0;i<n;i++) {
            parents.insert(dsu.find(i));
        }
        
        return parents.size();
    }
};

File: 0839-similar-string-groups/0839-similar-string-groups.java
class Solution {
    public int numSimilarGroups(String[] strs) 
    {
        int n=strs.length;
        
        int parent[]=new int[n];
        for(int i=0;i<n;i++)
        parent[i]=i;
        
        
        for(int i=0;i<n;i++)
        {
            String s1=strs[i];
            
            for(int j=i+1;j<n;j++)
            {
                String s2=strs[j];
                
                if(!isSimilar(s1,s2))
                continue;
                
                int p1=findParent(i,parent);
                int p2=findParent(j,parent);
                
                parent[p1]=p2;
            }
        }
        
        int ans=0;
        for(int i=0;i<n;i++)
        ans=parent[i]==i?ans+1:ans;
        
        return ans;
    }
    
    public boolean isSimilar(String s1,String s2)
    {
        if(s1.equals(s2))
        return true;
        
        int diff=0;
        for(int i=0;i<s1.length();i++)
        {
            if(s2.charAt(i)!=s1.charAt(i))
            diff++;
        }
        
        return diff<=2;
    }
    
    public int findParent(int i,int parent[])
    {
        if(parent[i]==i)
        return i;
        
        int p=findParent(parent[i],parent);
        parent[i]=p;
        
        return p;
    }
}